Feature: Command-line interface

davemfish opened the pull request:
This PR adds a command-line interface.

I also added `geometamaker.describe_dir` and `geometamaker.validate_dir` to more easily process a batch of files in a directory tree.

Fixes #43 

### --help
```
λ geometamaker --help
Usage: geometamaker [OPTIONS] COMMAND [ARGS]...

Options:
  -v         Override the default verbosity of logging. Use "-vvv" for debug-
             level logging. Omit this flag for default, info-level logging.
  --version  Show the version and exit.
  --help     Show this message and exit.

Commands:
  config    Configure GeoMetaMaker with information to apply to all metadata
            descriptions
  describe  Generate metadata for geospatial or tabular data, or zip archives.
  validate  Validate metadata documents for syntax or type errors.
```

```
λ geometamaker describe --help
Usage: geometamaker describe [OPTIONS] FILEPATH

  Describe properties of a dataset given by FILEPATH and write this metadata
  to a .yml sidecar file. Or if FILEPATH is a directory, describe all datasets
  within.

Options:
  -r, --recursive  if FILEPATH is a directory, describe files in all
                   subdirectories
  -nw, --no-write  Dump metadata to stdout instead of to a .yml file. This
                   option is ignored if `filepath` is a directory
  --help           Show this message and exit.
```

```
λ geometamaker validate --help
Usage: geometamaker validate [OPTIONS] FILEPATH

  Validate a .yml metadata document given by FILEPATH. Or if FILEPATH is a
  directory, validate all documents within.

Options:
  -r, --recursive  if `filepath` is a directory, validate documents in all
                   subdirectories.
  --help           Show this message and exit.
```

```
λ geometamaker config --help
Usage: geometamaker config [OPTIONS]

  When prompted, enter contact and data-license information that will be
  stored in a user profile. This information will automatically populate
  contact and license sections of any metadata described on your system. Press
  enter to leave any field blank.

Options:
  --individual-name TEXT
  --email TEXT
  --organization TEXT
  --position-name TEXT
  --license-title TEXT    the name of a data license, e.g. "CC-BY-4.0"
  --license-path TEXT     a url for a data license
  -p, --print             Print your current GeoMetaMaker configuration.
  --delete                Delete your configuration file.
  --help                  Show this message and exit.
```

### Some examples:
```
(C:\Users\dmf\projects\geometamaker\env) λ geometamaker describe data
01/06/2025 11:24:12  geometamaker.geometamaker INFO     data\data_no_csv.zip described
01/06/2025 11:24:12  geometamaker.geometamaker INFO     data\carbon_pools.csv described
01/06/2025 11:24:13  geometamaker.geometamaker WARNING  LOCAL_CS["Undefined SRS",LOCAL_DATUM["unknown",32767],UNIT["unknown",0],AXIS["Easting",EAST],AXIS["Northing",NORTH]] cannot be interpreted as a coordinate reference system
01/06/2025 11:24:13  geometamaker.geometamaker INFO     data\foo.gpkg described
01/06/2025 11:24:13  geometamaker.geometamaker INFO     data\data.zip described
```

```
(C:\Users\dmf\projects\geometamaker\env) λ geometamaker validate data
○ data\bad_config.yml does not appear to be a geometamaker document
✓ data\carbon_pools.csv.yml
✓ data\data.zip.yml
✓ data\data_no_csv.zip.yml
✕ data\DEM_gura.tif.yml: 2 validation errors
edition
    Input should be a valid string. [input_value=2, input_type=int]
keywords
    Input should be a valid list. [input_value=hello world keywords, input_type=str]
✓ data\foo.gpkg.yml
✕ data\watershed_gura.shp.yml: 1 validation errors
keywords
    Input should be a valid list. [input_value=foo bar, input_type=str]
```

phargogh commented on src/geometamaker/cli.py:
```diff
@@ -0,0 +1,155 @@
+        location = ', '.join(e['loc'])
+        msg_string = (f"    {e['msg']}. [input_value={e['input']}, "
+                      f"input_type={type(e['input']).__name__}]")
+        click.secho(location, bold=True)
+        click.secho(msg_string)
```
This is a _much_ nicer way to format text strings than the old terminal control codes!

phargogh commented on src/geometamaker/geometamaker.py:
```diff
@@ -400,3 +404,121 @@ def describe(source_dataset_path, profile=None):
+    return (yaml_files, messages)
+
+
+def describe_dir(directory, recursive=False):
```
Having just read through the CLI configuration, I totally get why this is the way it is, and I'm not sure I'd change it.

davemfish replied:
Yeah it's mostly in service of the CLI. And I thought the most conventional pattern there was the `-r` flag, rather having separate `describe` and `describe directory` commands. But I too am interested to see people using this and reacting to it!

phargogh commented on src/geometamaker/cli.py:
```diff
@@ -0,0 +1,155 @@
+    if not value or ctx.resilient_parsing:
+        return
+    config = geometamaker.Config()
+    click.confirm(
+        f'Are you sure you want to delete {config.config_path}?',
+        abort=True)
```
Oooo nice!

phargogh commented on src/geometamaker/cli.py:
```diff
@@ -0,0 +1,155 @@
+@click.option('--position-name', prompt=True, default='')
+@click.option('--license-title', prompt=True, default='',
+              help='the name of a data license, e.g. "CC-BY-4.0"')
+@click.option('--license-path', prompt=True, default='',
+              help='a url for a data license')
```
The parameter name `path` and the use of `url` in the help text make me wonder about how the text should be provided.  Should it be a URL?   A path to a local file?  A URI instead?  Maybe an example or two would help clarify?

davemfish replied:
yeah good point. I think a URL makes the most sense. `path` here comes from the DataPackage schema (https://specs.frictionlessdata.io/data-package/#licenses) We're sort of trying to mirror that where possible, for ease of future interoperability. But I think it makes sense to change the user-facing name to `--license-url` in our case.

phargogh commented on src/geometamaker/cli.py:
```diff
@@ -0,0 +1,155 @@
+
+import geometamaker
+
+root_logger = logging.getLogger()
+root_logger.setLevel(logging.DEBUG)
+handler = logging.StreamHandler(sys.stdout)
+formatter = logging.Formatter(
+    fmt='%(asctime)s %(name)-18s %(levelname)-8s %(message)s',
+    datefmt='%m/%d/%Y %H:%M:%S ')
+handler.setFormatter(formatter)
```
Super minor, but might I suggest using the all-caps reference names to clarify that these are module-level variables?

davemfish replied:
Definitely, thank you!

phargogh reviewed (commented):
Thanks, @davemfish !  This looks really exciting, and I was especially interested to see `click`'s features put to great use.  I don't have any blocking comments/suggestions, just a few minor things that could be addressed here in this PR or else in a future PR, so I'll just add this as a Comment review, but happy to switch to an Approve if you'd like!

davemfish commented on src/geometamaker/geometamaker.py:
```diff
@@ -400,3 +404,121 @@ def describe(source_dataset_path, profile=None):
+def describe_dir(directory, recursive=False):
+    """Describe all compatible datasets in the directory.
+
+    Take special care to only describe multifile datasets,
+    such as ESRI Shapefiles, one time.
```
Yeah, I agree. I didn't want to resort to such explicit, special handling of the shapefile parts. But it just seemed simplest and most reliable in the end. It avoids having to open the dataset here, before calling `describe`. And avoids the pitfall of walking the directory and opening one of the other files first, like the `dbf` or `shx` and then creating `foo.dbf.yml` instead of `foo.shp.yml`. But maybe there is a more general solution at some point.

phargogh replied:
Yep, definitely agree, this is for sure the most straightforward way to handle the common files we know about!  When it comes time to look into a more general solution, a possible candidate might be `gdal.Dataset.GetFileList`([docs](https://gdal.org/en/latest/api/python/raster_api.html#osgeo.gdal.Dataset.GetFileList)), which returns a list of filenames that GDAL thinks are a part of the dataset.

davemfish commented on src/geometamaker/cli.py:
```diff
@@ -0,0 +1,155 @@
+@click.option('-r', '--recursive', is_flag=True, default=False,
+              help='if FILEPATH is a directory, describe files '
+                   'in all subdirectories')
+@click.option('-nw', '--no-write', is_flag=True, default=False,
+              help='Dump metadata to stdout instead of to a .yml file. '
+                   'This option is ignored if `filepath` is a directory')
```
Good idea!

davemfish commented:
> Thanks, @davemfish ! This looks really exciting, and I was especially interested to see `click`'s features put to great use. I don't have any blocking comments/suggestions, just a few minor things that could be addressed here in this PR or else in a future PR, so I'll just add this as a Comment review, but happy to switch to an Approve if you'd like!

Thanks for the speedy review @phargogh ! I pushed a few changes and addressed all your comments.

phargogh reviewed (approved):
Sounds great, thanks @davemfish !
